pkg/coalescing/reconciler.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package coalescing
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"go.opentelemetry.io/otel/api/trace"
+	"go.opentelemetry.io/otel/label"
+	"sigs.k8s.io/cluster-api-provider-azure/util/cache/ttllru"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type (
+	// ReconcileCache uses and underlying time to live last recently used cache to track high frequency requests.
+	// A reconciler should call ShouldProcess to determine if the key has expired. If the key has expired, a zero value
+	// time.Time and true is returned. If the key has not expired, the expiration and false is returned. Upon successful
+	// reconciliation a reconciler should call Reconciled to update the cache expiry.
+	ReconcileCache struct {
+		lastSuccessfulReconciliationCache ttllru.PeekingCacher
+	}
+
+	// ReconcileCacher describes an interface for determining if a request should be reconciled through a call to
+	// ShouldProcess and if ok, reset the cool down through a call to Reconciled
+	ReconcileCacher interface {
+		ShouldProcess(key string) (expiration time.Time, ok bool)
+		Reconciled(key string)
+	}
+
+	// reconciler is the caching reconciler middleware that uses the cache or
+	reconciler struct {
+		upstream reconcile.Reconciler
+		cache    ReconcileCacher
+		log      logr.Logger
+	}
+)
+
+// NewRequestCache creates a new instance of a ReconcileCache given a specified window of expiration
```

Comment:
```suggestion
// NewRequestCache creates a new instance of a ReconcileCache given a specified window of expiration.
```

do we not have the exported func linter enabled?